### ISSUE TITLE
Should BCP code rtrim string data?

### ISSUE THREAD

**Issue by srutzky** [FreeTDS/freetds#81]
Hi there. I had meant to post this issue myself (I am the one who answered Neil's -- @neilmcguigan -- [question on StackOverflow](http://stackoverflow.com/questions/38984053/freebcp-unicode-data-is-odd-byte-size-for-column-should-be-even-byte-size/40392898#40392898) and found the bug) after doing a little more research, but Neil beat me to it ( #80) and Freddy -- @freddy77 -- was incredibly  efficient in fixing it (kudos on being so responsive :-).

What I was researching was the behavior of the .NET **SqlBulkCopy** Class as it should be equivalent behavior (at least with respect to "bcp" for Microsoft SQL Server). The reason I was researching this was that it occurred to me that the "problem" with this `rtrim` function was larger than it just breaking `NVARCHAR` data if the last character was one of the 1000 Code Points starting with **U+20xx**.

Prior to this fix, the behavior of **freebcp** was:

1. If data is `VARCHAR` / 8-bit Extended ASCII: trim trailing spaces
1. If data is `NVARCHAR` / UTF-16:
   1. If final Code Point is **U+2020**: remove both bytes of said Code Point as they are both `20`, and continue to checking the string (in which case it might remove another U+2020 entirely, or remove half of a U+20xx Code Point). Removing the entire Code Point won't produce an error, but it will silently corrupt the data.
   1. If final Code Point is **U+20xx**: remove 2nd byte of said Code Point as it will be `20` (i.e. space) due to byte-swapping Little Endian encoding, and exit (unless Code Point was U+2020). This produces an error (as was reported in this issue).

The net-effect of this behavior was to:

1. always trim trailing spaces from `VARCHAR` data
1. never actually trim any spaces from `NVARCHAR` data (due to the byte-swapped representation of U+0020 being `20 00` which ends with `00`)
1. occasionally fail due to removing half of a Code Point and/or silently corrupt the data by removing the entire Code Point.

For cases that fall into categories 1 and 2 directly above (I assume those represent the vast majority and that category 3 was somewhat rare), there is a behavior difference in that `VARCHAR` data is trimmed while `NVARCHAR` data is not.

With this fix, the net-effect of the behavior is merely to remove category 3 (the errors and/or corrupted data). While this is definitely an improvement over having the errors and/or corrupted data, the fact that `VARCHAR` is trimmed while `NVARCHAR` is not (now an intended behavior instead of a bug) seems inconsistent and non-intuitive. I would think that it should work the same for both types of string data.

With consistent behavior in mind, I would further question the practice of doing the trimming in the first place. Why is it being done? I actually consider the trimming behavior (even if modified to also work for `NVARCHAR` / UTF-16 data) to be a case of "data corruption" as those trailing spaces _could_ be intentional, valid data. It does not seem (to me, at least) appropriate for a tool at this level to make that determination for the end-user, at least not without a configuration option to enable/disable; a simple import tool _should_ import the data _as is_, and any "massaging" of it / transformation, should happen in the database, when such behavior is desired. And, coming full-circle now, the .NET [SqlBulkCopy Class](https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs) does not appear to do any trimming of strings. The [ConvertValue](https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs#L1250) method does call [SqlParameter.CoerceValue](https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlParameter.cs#L825), but there is no trimming done there either.

So, my recommendation is:

* IF there is no technical reason for doing the rtrim, then it should probably removed entirely. And if this is needed for Sybase, then add an additional check to the same `if` condition that was updated in fix #80 to be something along the lines of `&& server_type = SYBASE`.
* ELSE, if the desire is to keep the current behavior of trimming, then:
  1. make it consistent by also working with `NVARCHAR` / UTF-16 data (i.e. if type=NCHAR / NVARCHAR then test for `*p == '\0' && *p-1 ==  ' '`)
  1. add a configuration option to disable the trimming

**Comment by freddy77** [comment]
You are right, was going to write a follows up.
Actually beside the rtrim I'm more concern about the NULL/empty detection (a single NUL character is detected as empty string).
I think the trimming is here to handle CHAR/VARCHAR. CHAR is always space padded.

The original commit for rtrim is:
commit 3ce5d22d8878b4c153e3499918ca7737dfcf7373
Author: Brian Bruns <brian@bruns.com>
Date:   Sat May 25 00:33:48 2002 +0000

    Bill Thompson's bcp and datetime patches

Quite old! More than 14 years ago! Surely NVARCHAR was not there. And for some reason apply only to VARCHAR, not CHAR. Considering that this code is handling loading data from a file I supposed the trim allows to import back data saved from a CHAR into VARCHAR (for instance a CHAR(100) to a VARCHAR(50)). I think would be better to save a rtrim version of CHAR in the file instead. But I'd like to check what MS bcp utility do in this case. Can you do the check for me (export data with MS bcp either (N)CHAR and (N)VARCHAR where the string is smaller then column size) ?

**Comment by srutzky** [comment]
Hi again. Sorry for the prolonged delay.

I would doubt that this behavior is intended to account for the CHAR / VARCHAR difference in blank-padding for CHAR / NCHAR. For CHAR / NCHAR, those are always blank-padded anyway so it doesn't matter if the value being sent is trimmed (actually, in that case, it is wasted effort / cycles since the end result will not be trimmed). For VARCHAR / NVARCHAR, there could be valid blank spaces at the end that should remain and not be trimmed, and those will be lost if trimming those spaces.

The only reason I can think of to have the rtrim in there is to account for "bad" data in import files. But, if true, I think that is a bad reason. Data should be scrubbed intentionally by the import process, not by the data transport mechanism. At least, not in such a way that is a) not disclosed to the user, and b) not configurable (i.e. able to be disabled).

Yes,`NVARCHAR` and `NCHAR` were valid datatypes in SQL Server 2000. So I think Brian's update is just an oversight. It could have been Bill's mistake that Brian just copy / pasted, but still.

I did just test BCP as you requested and there is no trimming of CHAR columns when exporting. Here is what I did:

In SQL Server:

```
USE [tempdb];
-- DROP TABLE dbo.ExportTest;
CREATE TABLE dbo.ExportTest (Col1 CHAR(100), Col2 VARCHAR(100), Col3 INT);
INSERT INTO dbo.ExportTest (Col1, Col2, Col3) VALUES ('A', 'B', 99);
```

In a Command Prompt window:

    bcp tempdb.dbo.ExportTest OUT ExportOut.txt -T -S . -c
    type ExportOut.txt

Result of "type" command:

    A                                                                                                   	B	99
